bio/core/fastq.d

```diff
@@ -0,0 +1,151 @@
+/*
+   This file is part of BioD.
+   Copyright (C) 2016   George Githinji <biorelated@gmail.com>
+   Permission is hereby granted, free of charge, to any person obtaining a
+   copy of this software and associated documentation files (the "Software"),
+   to deal in the Software without restriction, including without limitation
+   the rights to use, copy, modify, merge, publish, distribute, sublicense,
+   and/or sell copies of the Software, and to permit persons to whom the
+   Software is furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+   The bio.core.fastq module is based from a question that I posted on forums.dlang.org.
+   I have bundled the answer as a module to support parsing fastq files with D.
+   Credit should go to Rikki Cattermole.
+ */
+
+module bio.core.fastq;
+
+struct FastqRecord {
+    const(char)[] id;
+    const(char)[] seq;
+    const(char)[] qual;
+
+    static auto read(const(char)[] from) {
+        struct Result {
+            private {
+                const(char)[] source;
+                FastqRecord value;
+                bool isEmpty;
+            }
+
+            this(const(char)[] source) {
+                this.source = source;
+                popFront;
+            }
+
+            @property {
+                FastqRecord front() {
+                    return value;
+                }
+
+                bool empty() {
+                    return isEmpty;
+                }
+            }
+
+            void popFront() {
+                import std.string : indexOf;
+
+                if (source is null) {
+                    isEmpty = true;
+                    return;
+                }
+
+                void tidyInput() {
+                    foreach(i, c; source) {
+                        switch(c) {
+                            case 0: .. case ' ':
+                                    break;
```

Comment:
So as far as I understand, this function skips invalid characters before beginning of each record. Why is this not an error condition instead?